### ISSUE TITLE
Comparison with tcomb

### ISSUE THREAD

**Issue by ctrlplusb** [ianstormtaylor/superstruct#15]
Hey @ianstormtaylor 👋

Thanks for your work on this!

Are you familiar with [`tcomb`](https://github.com/gcanti/tcomb)?  If so, would you mind raising any distinctions between `superstruct` and `tcomb`. 👍

**Comment by ianstormtaylor** [comment]
Wow... `tcomb` looks very well thought out! I somehow never found it when I was looking through all of the existing validation/schema/types libraries (and I looked through tons!). Let me get back to you on that, it may make Superstruct obsolete, but I need to try it out. 😄

**Comment by ianstormtaylor** [comment]
Hey @ctrlplusb thanks for opening this, and pointing me to `tcomb`. I've check it out in more detail now, and I think I can offer a comparison of the two, and explain how `superstruct` differs in a way that I think is still valuable.

The libraries are very similar in their architecture. (I actually noticed a few things `superstruct` could be doing better from reading the `tcomb` docs, so I'll implement those soon.) And they both come at it from a typing perspective, so the APIs are also very similar—which is actually a good thing I think, since it means we've arrived at a pretty solid place.

```js
const User = struct({
  id: 'number',
  name: 'string',
  is_admin: 'boolean?',
})
```
```js
const User = t.struct({
  id: t.Number,
  name: t.String,
  is_admin: t.maybe(t.Boolean),
})
```

---

The biggest difference in opinion that `tcomb` makes is that it uses actual object-oriented classes that you create instances of. Such its really more of a "model" library than a pure validation one. 

```js
const user = User({ id: 1, name: 'Jane Smith' })
```
```js
const user = User({ id: 1, name: 'Jane Smith' })
```

The APIs look exactly the same, but only because `tcomb` allows you to leave off the `new` keyword. But they're actually doing two different things. `superstruct` is returning a plain Javascript object representation of the data (with default values, and validated). Whereas `tcomb` is creating a new instance of a `User` class. It encourages adding methods to the user via the prototype, it ensures that the data is immutable, and some other things.

One of my goals for `superstruct` was for it to not impact your existing data, and to let you use pure Javascript objects. I wanted this because I'd like to use `superstruct` in APIs and in places where I might not want to opt-in to immutable data (and to the extra APIs that `tcomb` ships to have to deal with it), and I want to preserve interoperability with all libraries that just take plain objects.

For an example of the decisions this opts you in to, updating data with `tcomb` looks something like:

```js
const p = Point({ x: 1, y: 2 });

p = Point.update(p, {x: { '$set': 3 }}); // => { x: 3, y: 2 }
```

Whereas with `superstruct` you're just dealing with plain objects.

---

The other main difference is that `superstruct` is built with typing as a guide, but with validation as one of its core goals. This means that the `tcomb-validate` library is also handled by `superstruct`, with what I think is a more first-class API. 

I debated whether this was a coupling of concerns, but because the typing use case is still handled gracefully with the `{ struct }` import, and it needs to throw `TypeErrors` anyways, I decided it didn't merit separating the two.

What this means is that as a user you get a more expressive API for validating out of the box, for things like API inputs, forms, etc.

---

And then the final difference, that is the most obvious when you look at the examples, is that `superstruct` offers a string-based API, instead of an object-based one, which I've found to be a lot easier to read and reason about. Instead of passing around objects for your types, you can refer to the scalar types as strings and define them in a single, common place. 

```js
const User = struct({
  id: 'number',
  name: 'string',
  is_admin: 'boolean?',
})
```
```js
const User = t.struct({
  id: t.Number,
  name: t.String,
  is_admin: t.maybe(t.Boolean),
})
```

For simple examples the object-driven approach reads easily as well, but once you get into more complex validations for APIs I've found that it gets hard to parse pretty quickly—especially when lots of optional fields are involved. However the object approach makes it easier to define one-off types (what `tcomb` calls "refinements" I believe) if you happen to need something a little different.

`superstruct` tries to borrow from the syntaxes of TypeScript, Go, GraphQL, etc. in this regard in terms of making it super-terse to define a struct. And it encourages you not to define arbitrary types without thinking them through and registering them beforehand in a common place.

---

So, all that said, they're very similar. And I like a lot of the decisions that `tcomb` has made, it seems very well thought out for its goals.

If you want a full-fledged "model" concept for your codebase, and you want immutability, it sounds like a good choice. If you're looking for a lighter-weight choice for validating plain objects, arrays, etc. then `superstruct` might be the better option. After doing more research, I think they both have a place. And I'll definitely be striving to crib some more of the good decisions from `tcomb`.

Hope that helps!

**Comment by ctrlplusb** [comment]
Amazing reply @ianstormtaylor 👍

Thanks so much for taking the time to answer this issue with such a thoughtful investigation. I have no doubt many others will find this answer helpful upon review.

Keep up the great work. 👍👍
